File: python/203 - Remove Linked List Elements.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeElements(self, head, val):
        """
        :type head: ListNode
        :type val: int
        :rtype: ListNode
        """
        # empty list
        if not head:
            return head
        
        it = head
        prev = head
        while it:
            if it.val == val:
                if it == head:
                    head = head.next
                    prev = it
                    it = it.next
                elif not it.next:
                    prev.next = None
                    break
                else:
                    prev.next = it.next
                    it = it.next
            else:
                prev = it
                it = it.next
        return head

File: cpp/203 - Remove Linked List Elements.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        // prev
        ListNode* new_head = new ListNode(0);
        new_head->next = head;
        
        // var
        ListNode* it = new_head;
        ListNode* prev = it;
        
        it = it->next;
        while (it) {
            if (it->val == val)
                prev->next = it->next;
            else
                prev = it;
            it = it->next;
        }
        return new_head->next;
    }
};